### ISSUE TITLE
Intermittant "uninitialized constant RunCaseJob" error when triggering job

### ISSUE THREAD

**Issue by epugh** [rails/solid_queue#474]
Periodically I get this error:
<img width="584" alt="image" src="https://github.com/user-attachments/assets/d41649c7-4198-46bb-9c89-e33175fc8e94" />

Here is the raw data:

```json
  {
  "id": 268,
  "queue_name": "bulk_processing",
  "class_name": "RunCaseJob",
  "arguments": {
    "job_class": "RunCaseJob",
    "job_id": "9fcfc846-a982-4ac8-b2e1-11878b2d97af",
    "provider_job_id": null,
    "queue_name": "bulk_processing",
    "priority": null,
    "arguments": [
      {
        "_aj_globalid": "gid://quepid/Case/6789"
      },
      {
        "_aj_globalid": "gid://quepid/Try/63146"
      }
    ],
    "executions": 0,
    "exception_executions": {},
    "locale": "en",
    "timezone": "UTC",
    "enqueued_at": "2024-12-30T16:08:10.153536324Z",
    "scheduled_at": "2024-12-30T16:08:10.150847860Z"
  },
  "priority": 0,
  "active_job_id": "9fcfc846-a982-4ac8-b2e1-11878b2d97af",
  "scheduled_at": "2024-12-30T16:08:10.150Z",
  "finished_at": null,
  "concurrency_key": null,
  "created_at": "2024-12-30T16:08:10.410Z",
  "updated_at": "2024-12-30T16:08:10.410Z"
}
```

If I retry the job (using mission control) it will give the same error.  However, if I resubmit a new RunCaseJob with the same arguments, then that second one runs.   

If you are curious, the PR that adds solid_queue to my open source app is here: https://github.com/o19s/quepid/pull/1039

Thanks for any ideas!

**Comment by Kaakati** [comment]
Same happening with me, I have an OTPMailer that SolidQueue says method not found, and another Job that says uninitialized constant.

**Comment by epugh** [comment]
It seems to happen when I directly use `perform_later`, however I have a [nightly recurring](https://github.com/o19s/quepid/pull/1039/files#diff-e02d060e76a92603bab92088890870d4c859fb237e9ccc4283766ca579acdddb) job that doesn't seem to have the same issue.    I'm wondering if it's some sort of order of loading classes?

**Comment by rosa** [comment]
Hey @epugh, thanks for the link to the app and the PR. That's very strange! I wonder if, given that this branch hasn't been merged yet, is possible you're enqueuing that job and it's being picked up by a worker that doesn't have that branch's code? 

@Kaakati, could you provide more details about the errors you're getting? With backtraces if possible, and the lines of code that might be relevant.

**Comment by jbennett** [comment]
Hmm, I am seeing this too in a new Rails 8 app. This works in dev but gives a name error in staging/production. Similar to @epugh this works for `perform_now` but not for `perform_later`. I'm calling my job with: `MyJob.perform_later(active_record_object)`

The error in solid queue is:

```json
{
  "exception_class": "NameError",
  "message": "uninitialized constant MyJob",
  "backtrace": [
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/inflector/methods.rb:290:in `constantize'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/core_ext/string/inflections.rb:74:in `constantize'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/core.rb:63:in `deserialize'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:28:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:120:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/railtie.rb:95:in `block (4 levels) in <class:Railtie>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/reloader.rb:77:in `block in wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/execution_wrapper.rb:87:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/reloader.rb:74:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/railtie.rb:94:in `block (3 levels) in <class:Railtie>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `instance_exec'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:140:in `run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:27:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/models/solid_queue/claimed_execution.rb:95:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/models/solid_queue/claimed_execution.rb:61:in `perform'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/pool.rb:23:in `block (2 levels) in post'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/execution_wrapper.rb:91:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/app_executor.rb:7:in `wrap_in_app_executor'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/pool.rb:22:in `block in post'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:24:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `block in synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:22:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/ivar.rb:170:in `safe_execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/future.rb:55:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:359:in `run_task'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:350:in `block (3 levels) in create_worker'",
    "<internal:kernel>:187:in `loop'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:341:in `block (2 levels) in create_worker'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `catch'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `block in create_worker'"
  ]
}
```

I am also sometimes getting an error from a recurring job:

```json
{
  "exception_class": "NameError",
  "message": "uninitialized constant SolidQueue::RecurringJob::Legacy",
  "backtrace": [
    "(eval at /usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/jobs/solid_queue/recurring_job.rb:7):1:in `perform'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/jobs/solid_queue/recurring_job.rb:7:in `eval'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/jobs/solid_queue/recurring_job.rb:7:in `perform'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:68:in `block in _perform_job'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:120:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/i18n-1.14.6/lib/i18n.rb:353:in `with_locale'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/translation.rb:9:in `block (2 levels) in <module:Translation>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `instance_exec'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/core_ext/time/zones.rb:65:in `use_zone'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/timezones.rb:9:in `block (2 levels) in <module:Timezones>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `instance_exec'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:140:in `run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:67:in `_perform_job'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/instrumentation.rb:32:in `_perform_job'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:51:in `perform_now'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/instrumentation.rb:26:in `block in perform_now'",
    "/usr/local/bundle/ruby/3.3.0/gems/activerecord-8.0.0/lib/active_record/railties/job_runtime.rb:13:in `block in instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/instrumentation.rb:40:in `block in instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/notifications.rb:210:in `block in instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/notifications/instrumenter.rb:58:in `instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/notifications.rb:210:in `instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/instrumentation.rb:39:in `instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activerecord-8.0.0/lib/active_record/railties/job_runtime.rb:11:in `instrument'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/instrumentation.rb:26:in `perform_now'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/logging.rb:32:in `block in perform_now'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/tagged_logging.rb:143:in `block in tagged'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/tagged_logging.rb:38:in `tagged'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/tagged_logging.rb:143:in `tagged'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/broadcast_logger.rb:241:in `method_missing'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/logging.rb:39:in `tag_logger'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/logging.rb:32:in `perform_now'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:29:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:120:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/railtie.rb:95:in `block (4 levels) in <class:Railtie>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/reloader.rb:77:in `block in wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/execution_wrapper.rb:87:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/reloader.rb:74:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/railtie.rb:94:in `block (3 levels) in <class:Railtie>'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `instance_exec'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:129:in `block in run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/callbacks.rb:140:in `run_callbacks'",
    "/usr/local/bundle/ruby/3.3.0/gems/activejob-8.0.0/lib/active_job/execution.rb:27:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/models/solid_queue/claimed_execution.rb:95:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/app/models/solid_queue/claimed_execution.rb:61:in `perform'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/pool.rb:23:in `block (2 levels) in post'",
    "/usr/local/bundle/ruby/3.3.0/gems/activesupport-8.0.0/lib/active_support/execution_wrapper.rb:91:in `wrap'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/app_executor.rb:7:in `wrap_in_app_executor'",
    "/usr/local/bundle/ruby/3.3.0/gems/solid_queue-1.1.0/lib/solid_queue/pool.rb:22:in `block in post'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:24:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `block in synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/synchronization/mutex_lockable_object.rb:48:in `synchronize'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/safe_task_executor.rb:22:in `execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/ivar.rb:170:in `safe_execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/future.rb:55:in `block in execute'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:359:in `run_task'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:350:in `block (3 levels) in create_worker'",
    "<internal:kernel>:187:in `loop'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:341:in `block (2 levels) in create_worker'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `catch'",
    "/usr/local/bundle/ruby/3.3.0/gems/concurrent-ruby-1.3.4/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:340:in `block in create_worker'"
  ]
}
```

**Comment by rosa** [comment]
@jbennett, does it happen consistently in staging/production for `MyJob`? Where is `MyJob` defined? As for the recurring job, could you copy the entry for that in `recurring.yml`?

**Comment by jbennett** [comment]
This looks to be happening consistently. The job file is located at `app/jobs/update_coordinate_job.rb` with `class UpdateCoordinateJob < ApplicationJob`.

The recurring job is:

```yml
production:
  daily_sync:
    command: Legacy.import_all
    schedule: every day at 3am
```

I did try changing the command to `::Legacy.import_all` which changed the error to `uninitialized constant Legacy`. It's almost like SolidQueue doesn't have access to any of the application classes

**Comment by rosa** [comment]
Huh, that's so odd! How are you deploying the code to production and how are you starting solid queue?

**Comment by jbennett** [comment]
This is a staging server but it's running with a production environment/configuration. It using using a single server setup with the solid queue in puma plugin with the environment variable for that turned on.

**Comment by rosa** [comment]
Oh, interesting! @jbennett, can you reproduce this if you run Solid Queue via `bin/jobs` instead of via the Puma plugin? And with the Puma plugin, are you [preloading your app](https://nts.strzibny.name/preloading-applications-in-production/)?

**Comment by epugh** [comment]
I wanted to follow up as I gave my team the demo of the branch that uses Solid Queue, and looking forward to merging and deploying it once we get past this issue.

I checked and my [puma.rb](https://github.com/o19s/quepid/blob/judge_judy/config/puma.rb) doesn't specify preloading, so I believe it is as the default as I am on [puma 6.5](https://github.com/o19s/quepid/blob/judge_judy/Gemfile.lock#L359).  

My set up is also *NOT* running solid queue in the puma plugin.

The staging version of Quepid is running on Heroku, and the defined start script was `bundle exec rake solid_queue:start`.  I am going to change that to use the [bin/jobs](https://github.com/o19s/quepid/blob/judge_judy/app.json#L86) instead to see what happens.   

If you are curious @rosa and want an account on the staging server to see "Mission Control" (it's all public/opensource) I can send you a link.

**Comment by rosa** [comment]
@epugh, does the [RunJudgeJudyJob](https://github.com/o19s/quepid/blob/judge_judy/app/jobs/run_judge_judy_job.rb) fail consistently? Or is it just sometimes? 

> If you are curious @rosa and want an account on the staging server to see "Mission Control" (it's all public/opensource) I can send you a link.

If it fails consistently, maybe that helps debugging 🤔  I haven't been able to reproduce this.

**Comment by jbennett** [comment]
I deploy to the production environment a couple days ago and it has been reliable. Only difference I can really see is running separately via bin/jobs vs in puma on staging.

I'm seeing anything to avoid preloading or eager loading. Anything else I could be looking for in that direction?

**Comment by epugh** [comment]
Likewise is that I've been using bin/jobs recently...     And things have been reliable...   I did also set all the settings back to the defaults..    

I'll keep you posted on this!

**Comment by jbennett** [comment]
Unfortunately I need to take back the "reliable in production comment". This is up and running in production but I am seeing the same name errors in production and staging. I'm setting `config.eager_load=true` and `config.rake_eager_load=true`. If I `rails c` and do a `SolidQueue::Job.find(123).dispatch` it will run but the process kicked off by `bin/jobs` is pretty consistently giving the name error.

I was originally using the puma plugin for staging but switched to bin/dev to try and work this out. Either way has been causing issues

**Comment by jbennett** [comment]
TLDR: This was an issue on my end and SolidQueue is currently running beautifully with the Puma plugin in staging and with `bin/jobs` in production. The issue was an old container running a 2 month old version of the app.

---

I was originally using Kamal with a single destination, but split it into `deploy.staging.yml` and `deploy.prod.yml`. This worked great until I started using jobs. After far to much chasing down wrong paths, I noticed an odd container when running `docker container ls`:

```bash
9fe40e3246c3 ... PROJECT-web-staging
a776847b129c ... PROJECT-job-staging
8e171eede8ed ... PROJECT-web
```

🤦🏻‍♂️

`docker stop 8e171eede8ed` killed the pre-destination-split web server that was dutifully trying to process jobs despite being 2 months out-of-date, and not having the correct gems and code.

This also explains why TurboBroadcast jobs were running fine (gem installed early) but Noticed was not.

**Comment by kaka-ruto** [comment]
Incase anyone is still having this issue, I figured out my problem and solution here - https://github.com/rails/solid_queue/issues/276#issuecomment-2762916303

**Comment by josephecombs** [comment]
> Incase anyone is still having this issue, I figured out my problem and solution here - [#276 (comment)](https://github.com/rails/solid_queue/issues/276#issuecomment-2762916303)

Just wasted a solid half a day debugging this. This was the exact issue. A stale, old container was still alive and intermittently processing jobs from sqlite. 

**_THE REAL ISSUE IS THAT KAMAL LETS STALE CONTAINERS KEEP LIVING AND PROCESSING JOBS IN PROD!!!! THAT SEEMS INSANE!!!!_**

**Comment by rosa** [comment]
Huh, @josephecombs, that seems like something is not working quite right. Kamal should replace running containers with new ones on a deploy. Have you opened an issue in the Kamal repo with detailed info about your system (versions, config) and reproduction steps?

**Comment by jbennett** [comment]
When this happened to me I had renamed my destination from production to prod which is why the container was still around if I am not mistaken.
